GDB 12

bsdjhb opened the pull request:
CHERI GDB 12 changes the way it deals with tags.  This updates the register descriptions and register transfers to store tags as part of each cap register and removing the tag bitmask virtual register.

bsdjhb committed:
Switch to the Morello cap register tdesc from Morello GDB.

In particular, this uses a new capability register type, drops
the read-only TLS register, and adds CCTLR.

bsdjhb committed:
Return the tag in a leading byte for capability registers.

bsdjhb committed:
Morello: Update for CHERI GDB 12.

This drops the tag_map and stores tags inline for capability registers.

bsdjhb committed:
Export ctpidr_el0, rcsp, rddc, rctpidr_el0, and cid_el0 to gdb.

bsdjhb committed:
Fail writes to cctlr and treat it as a read-only register.

bsdjhb committed:
Update CHERI-RISC-V to work with CHERI GDB 12.

arichardson commented on target/arm/helper.c:
```diff
@@ -428,29 +411,21 @@ static int aarch64_gdb_set_cheri_reg(CPUARMState *env, uint8_t *mem_buf, int n)
+        return CHERI_CAP_SIZE + 1;
     case 32:
-        return CHERI_CAP_SIZE;
+        return CHERI_CAP_SIZE + 1;
```
Could also remove those returns and just have a single one?

arichardson reviewed (approved):
Not tested this, but looks good to me.

arichardson commented:
clang-format might make checkpatch.pl happy.

brooksdavis commented:
This isn't working for me at all. When I connect using the command that from `cheribuild --wait-for-debugger ` that use to work I get:
```
Remote debugging using localhost:1234
Truncated register 32 in remote 'g' packet
```
qemu tree is at 8078d61697f87cc1384d6ef24b3e1da36d0b7cde
gdb tree is at CTSRD-CHERI/gdb@ddd1b56a1dcec42f549b140dfeee394c93599072

bsdjhb commented:
Hmm, GDB 12 is working for me against both RISC-V and Morello.  For RISC-V it fails to set the architecture correctly from the XML sent by QEMU so I have to load a file first (like the kernel), but for Morello it auto-selects the right architecture even without a symbol file loaded.  Did you have any earlier messages?  When I did trigger a similar error on RISC-V without a file loaded there were earlier errors:

```
(gdb) target remote localhost:1234
Remote debugging using localhost:1234
warning: Architecture rejected target-supplied description
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
Truncated register 37 in remote 'g' packet
(gdb) show architecture 
The target architecture is set to "auto" (currently "i386").
```

(Specifically, the "Architecture rejected target-supplied description" warning)

bsdjhb commented:
The issue is with purecap binaries as GDB treats the "code_ptr" for the "pc" register as being a capability when the main symbol file is purecap.  I have a workaround for GDB to ignore code_ptr/data_ptr XML register types if the bitsize doesn't match the default size of pointers.

bsdjhb committed:
Support for fetching in-memory capabilities via GDB.

qXfer:capa:read reads a single capability.  The annex portion of the
command is the virtual address of the capability.  The offset and
length fields are relative to the 17 byte returned value.  The tag is
returned in the first byte.

bsdjhb commented:
Fixed the tabs and also added a commit that reads in-memory capabilities for system mode.

bsdjhb commented:
I've fixed some of the clang format warnings, but clang-format wants to reformat a bunch of existing nearby lines, and in some cases warns about code that as just duplicated from similar functions e.g. flatview_readcap() vs flatview_read() both using the same indentation.

arichardson commented:
The remaining clang-format suggestions can be ignored. The ones in gdbstub.c should not be applied, whereas for the other ones I don't mind either way.
